pytensor/compile/builders.py

```diff
@@ -469,327 +416,254 @@ def __str__(self):
-            all_grads_l = []
-            all_grads_ov_l = []
-            for inp, fn_gov in zip(local_inputs, goverrides_l):
-                if fn_gov == "default":
-                    gnext, gnext_ov = OpFromGraph._filter_grad_var(next(gdefaults), inp)
-                    all_grads_l.append(gnext)
-                    all_grads_ov_l.append(gnext_ov)
-                elif isinstance(fn_gov, Variable):
-                    if isinstance(fn_gov.type, DisconnectedType | NullType):
-                        all_grads_l.append(inp.zeros_like())
-                        all_grads_ov_l.append(fn_gov.type())
-                    else:
-                        raise ValueError(self.STYPE_ERR_MSG % fn_gov.type)
-                else:
-                    if not callable(fn_gov):
-                        raise TypeError(self.TYPE_ERR_MSG % fn_gov)
-                    gov, gov_ov = OpFromGraph._filter_grad_var(
-                        fn_gov(*callable_args), inp
-                    )
-                    all_grads_l.append(gov)
-                    all_grads_ov_l.append(gov_ov)
+            default_input_grads = fn_grad(wrt=wrt) if wrt else []
+            input_grads = self._combine_list_overrides(
+                default_input_grads, custom_input_grads, callable_args
+            )
         else:
-            # callable case
-            if not callable(lop_op):
-                raise TypeError(self.TYPE_ERR_MSG % lop_op)
-            goverrides_l = lop_op(*callable_args)
-            if not isinstance(goverrides_l, list):
-                raise TypeError(
-                    "Gradient/L_op overriding function should return a list, "
-                    f'got "{type(goverrides_l)}"'
-                )
-            all_grads_l, all_grads_ov_l = zip(
-                *[
-                    OpFromGraph._filter_grad_var(grad, inp)
-                    for grad, inp in zip(goverrides_l, local_inputs)
-                ]
+            input_grads = self._call_custom_override(
+                lop_op_constructor, callable_args, nin
             )
-            if len(all_grads_l) != len(local_inputs):
-                raise ValueError(
-                    "Gradient/L_op overriding function should return list of "
-                    f"{int(inp_len)} outputs, got {len(all_grads_l)}"
-                )
-        all_grads_l = list(all_grads_l)
-        all_grads_ov_l = list(all_grads_ov_l)
-        self._lop_op = type(self)(
-            inputs=local_inputs + local_outputs + output_grads,
-            outputs=all_grads_l,
+
+        # Filter out disconnected input and output gradients
+        connected_input_grads = [
+            inp_grad
+            for inp_grad in input_grads
+            if not isinstance(inp_grad.type, DisconnectedType | NullType)
+        ]
+        connected_output_grads = [
+            out_grad
+            for out_grad in output_grads
+            if not isinstance(out_grad.type, DisconnectedType)
```

Comment:
Why don't output_grads need to check for `NullType`?

Reply:
Honestly because I am not sure when Nulltype actually arise. 

I prefer the special logic to be as specific as possible, we can reassess if NullTypes also show up in the future?

Reply:
But the same logic should apply yeah, let's put Null here as well

Reply:
Actually I don't want to filter those out. If we know that output gradient is Null, we shouldn't omit that information from the inner gradient graph generation. If we omit `grad` assumes it's simply disconnected.

Omitting disconnected makes sense, because they will default again to disconnected_types if needed by `grad`